[com_finder] define identifier field for smart search plugins

miguelbgouveia opened the pull request:
Pull Request for Issue # .
#### Summary of Changes
#### Testing Instructions

Creation of an identifier field property in order to define the table field that store the identifier.
In order to create a plugin to components that have identifier field in table other than id.

miguelbgouveia committed:
Define identifier field for smart search plugins

Creation of an identifier field property in order to define the table field that store the identifier.
In order to create a plugin to components that have identifier field in table other than id.

miguelbgouveia commented:
I was creating an smart search plugin for a specific component.
The problem is that component use a table where the identifier field don't have the 'id' name.
To resolve the problem it will be good if I could specified the field identifier name in cases that the name is other than 'id'.

miguelbgouveia committed:
Correction and alignment of comments 

To pass in the CI tool.

miguelbgouveia committed:
Another documentation space correction.

To pass in the CI rules

miguelbgouveia committed:
Correction of error in checkItemAccess function

Error in given the table alias when it shouldn't.

miguelbgouveia committed:
Merge branch 'staging' into patch-2

roland-d commented on administrator/components/com_finder/helpers/indexer/adapter.php:
```diff
@@ -927,4 +937,29 @@ protected function translateState($item, $category = null)
+			$tableAndIdentifier = $tableAlias . '.' . $this->identifier_field;
+		}
+
+		$query->where($tableAndIdentifier . ' = ' . (int) $id);
```
I think it is better to return the $query rather than by reference so it is clear what is going on.

miguelbgouveia replied:
You right, I will actualize the code with your suggestion,

andrepereiradasilva commented on administrator/components/com_finder/helpers/indexer/adapter.php:
```diff
@@ -927,4 +937,29 @@ protected function translateState($item, $category = null)
+	 *
+	 * @return  void
+	 *
+	 * @since   3.5
```
while you are at it replace all `@since` tags by `@since   __DEPLOY_VERSION__`

obvisualy this is not going to be merged on 3.5 and this way you don't need to change them again.

miguelbgouveia committed:
Return the query in filterByIdentifier 

Returns the query with the identifier filter instead of change it by reference. Just for better readability.

andrepereiradasilva commented on administrator/components/com_finder/helpers/indexer/adapter.php:
```diff
@@ -121,6 +121,14 @@
+	 * The field the identifier is stored in.
+	 *
+	 * @var    string
+	 * @since  3.5
```
while you are at it replace all `@since` tags by `@since __DEPLOY_VERSION__`

obvisualy this is not going to be merged on 3.5 and this way you don't need to change them again.

miguelbgouveia committed:
Change since version

Using the dynamic tag __DEPLOY_VERSION__ instead of a fixed version.

brianteeman commented:
I am tempted to close this. It is solving a problem that shouldnt exist as you should have an id

@chrisdavenport thoughts<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9212">issues.joomla.org/joomla-cms/9212</a>.</sub>

mbabker commented:
With this patch, it would be consistent with JTable which has API methods to allow you to define a primary key other than an `id` column.  Also note the core database schema does not always use `id` as the primary key; the `#__extensions` table uses a `extension_id` primary key.

brianteeman commented:
@mbabker well can you test it then!!

chrisdavenport commented:
I've come across this problem before, but I've always been able to solve it by renaming the column.  Others may not be so fortunate.  So I think it makes sense and we should accept it once we have the usual 2+ independent tests.

euismod2336 commented:
I have tested this item :red_circle: unsuccessfully on f13f36b2f5c0a1c27ca36ef958d5d01c22665f15<br /><br />Created test table with `pk` as primary key. Created smart search plugin to index the table. Indexing went correct if `$item->id = $item->pk` and incorrect if `$item->id` was not set. After clearing and applying patch, modified the identifier_field but items not indexed.

I think the change isn't thorough enough, in functions like `categoryStateChange`, where all the children are modified, it should also account for the new field.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9212">issues.joomla.org/tracker/joomla-cms/9212</a>.</sub>

miguelbgouveia commented:
My PR should work, can you send me your smart seach plugin to see what goes wrong?

And you are right, in the categoryStateChange should consider this field as well. I will try to correct that.

euismod2336 commented:
@miguelbgouveia http://paste.ofcode.org/iRz7rxGMESsH2MP2z9xDAU at the bottom i added the sql from the table i used.

franz-wohlkoenig commented:
@miguelbgouveia is there something new to test?

miguelbgouveia commented:
Unfortunately I am not more working with the joomla CMS and and I have no time to verify and correct the problem in the present. Maybe in future I can attend this issue.

mbabker commented:
This PR needs a new owner or it will have to be closed as abandoned.

brianteeman commented:
As it has been several months since the last comment by @mbabker I am closing this as abandoned
